utils/char16trie/README.md

```diff
@@ -0,0 +1,44 @@
+# icu_char16trie [![crates.io](https://img.shields.io/crates/v/icu_char16trie)](https://crates.io/crates/icu_char16trie)
+
+## icu_char16trie [![crates.io](http://meritbadge.herokuapp.com/icu_char16trie)](https://crates.io/crates/icu_char16trie)
+
+`icu_char16trie` is a utility crate of the [`ICU4X`] project.
+
+This component provides a data structure for an time-efficient lookup of values
+associated to code points.
+
+It is an implementation of the existing [ICU4C UCharsTrie](https://unicode-org.github.io/icu-docs/apidoc/released/icu4c/classicu_1_1UCharsTrie.html)
+/ [ICU4J CharsTrie](https://unicode-org.github.io/icu-docs/apidoc/released/icu4j/com/ibm/icu/util/CharsTrie.html) API.
+
+### Architecture
+
+ICU4X [`Char16Trie`](crate::char16trie::Char16Trie) is designed to provide a read-only view of UCharsTrie data that is exported from ICU4C.
+
+### Examples
+
+#### Querying a `Char16Trie`
+
+```rust
+use icu_char16trie::char16trie::{Char16Trie, Char16TrieIterator, TrieResult};
+use zerovec::ZeroVec;
+
+// A Char16Trie containing the ASCII characters 'a' and 'b'.
+let trie_data = vec![48, 97, 176, 98, 32868];
+let trie = Char16Trie {
+    data: ZeroVec::from_slice(trie_data.as_slice()),
```

Comment:
FYI  Henri asked for a more direct way to create this object from an array slice, or really also a more direct way to create what's currently called the Char16TrieIterator from an array slice.

If it's not necessary to make a named type for what's really just a slice of an array or of a ZeroVec, then it would be nice to cut out this layer.

Reply:
@sffc Thoughts on this? Like Markus says, there's no real use for a separate struct from the iterator right now, but maybe you had something in mind for the future?

Data coming from a provider will most likely already be a ZeroVec, and there's no need to create a Char16Trie first, it's possible to create the Char16TrieIterator directly from a slice of ZeroVec data.

Reply:
As in other areas, I tend to prefer the extra wrapper layer; it is a zero-cost abstraction that makes code a little bit more readable and maintainable.  For example, I prefer this:

```rust
let trie: Char16Trie = // get trie from data provider
let mut iter = trie.iter();
```

over this:

```rust
let zv: ZeroVec<u16> = // get zerovec from data provider
let mut iter: Char16TrieIterator = zv.as_slice().into()
```

Reply:
In the data provider itself, I also prefer the readability of the extra type:

```rust
#[icu_provider::data_struct]
pub struct MyDataStruct<'data> {
    /// A map from strings to values
    trie: Char16Trie<'data>,
}
```

Otherwise, you have this:

```rust
#[icu_provider::data_struct]
pub struct MyDataStruct<'data> {
    /// A vector of numbers that represent a Char16Trie, a map from strings to values
    trie_data: ZeroVec<'data, u16>,
}
```

That being said, the extra type may not be a zero-cost abstraction in serde; but if this is a problem, we can add `#[serde(flatten)]` to make it zero-cost.

Reply:
My impression was that for the current use cases (collator, segmenter) the trie data is interleaved with other data at unpredictable offsets, so I think we'll end up constructing the iterator using a slice, rather than using the `Char16Trie` structure directly. But I don't see any harm in keeping `Char16Trie`, it might be more useful in the future, and I might be wrong about the current use cases.